### PR TITLE
Store Search-API-relative URL for all saved searches.

### PR DESCRIPTION
**Pull request by galund**
As well as saving some bytes, and making our saved search data portable between environments, this removes a theoretical attack vector i.e. where a client sends a malicious URL into the API as a 'saved search'.

https://trello.com/c/jjVWnK1e/722-security-fixes-for-save-search-feature

### CODE REVIEW

File: app/url_utils.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    # convert args to str; urllib can accept bytes but our string ops can't
+    base_url, url = str(base_url), str(url)
+
+    # remove the netloc and port from the url...
+    parsed_url = urlsplit(url)
+    sanitised_url = urlunsplit(('', '', parsed_url.path, parsed_url.query, parsed_url.fragment))
+
+    # ...and remove any common path prefix up to and including the final slash
+    base_path = urlsplit(base_url).path or '/'
+    base_path = base_path[:base_path.rfind('/') + 1]
```
Comment (benvand) [inline app/url_utils.py]: Why do we care about the preceding slash? why not always have it or always don't have it?
Also why do we care about the matching bits? Can't we just store this as 
```
url = urlsplit(str(url))
urlunsplit(['', '', url.path, url.query, url.fragment])
```

File: app/url_utils.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from urllib.parse import urlsplit, urlunsplit
```
Comment (benvand) [inline app/url_utils.py]: I'd drop this in the multi purpose utils file. Seems overkill to have another file for a single function.

File: app/url_utils.py (hunk 1 above)
Comment (galund) [reply app/url_utils.py]: So it seemed weird to have a field where the first character was _always_ a slash.

So then I wanted a consistent/logical process for getting rid of that first slash, and I wanted something that could be expressed in terms of an explainable function. So this is really to cover some edge cases that shouldn't ever be seen in real life, but are (somewhat?) explained by the tests.

I'll admit that those lines only got added after I'd factored the code into a separate function and given it a name, and then thought about the implications of the edge cases. They aren't really essential to what we're doing, but making this a function that will behave consistently with its 'reverse' i.e. `urljoin` seemed like a nice idea.

File: app/url_utils.py (hunk 2 above)
Comment (galund) [reply app/url_utils.py]: Personally I think multi-purpose utils files are the spawn of the devil. They only ever get longer and longer and longer and before you know it you have a 4000 line pile of spew. I've seen it (in a previous employment that had existed for longer than GDS, admittedly).

**Comment by galund** [comment]
For added clarity: the weird prefix-removal is necessary in order for the last two test cases to pass, neither of which we expect to happen in real life in our application, but I think are necessary if the function is to have well-defined behaviour in all the edge cases. (Including the case of our search API being mounted somewhere other than `/`, which seems like a terribly bad idea but not entirely impossible.)

File: app/url_utils.py (hunk 2 above)
Comment (benvand) [reply app/url_utils.py]: Unless we're planning on splitting out that other file into different files, ie `json_utils`, `request_utils`, `data_utils` then an extra file is inconsistent. :man_shrugging:

File: app/url_utils.py (hunk 1 above)
Comment (benvand) [reply app/url_utils.py]: It only removes the preceding slash in some cases though. That's the bit that I don't get.

File: app/url_utils.py (hunk 2 above)
Comment (samuelhwilliams) [reply app/url_utils.py]: Got to start somewhere.

File: app/url_utils.py (hunk 1 above)
Comment (galund) [reply app/url_utils.py]: Yup, this is quite hard to explain. The case where we don't do anything at all is where there isn't any kind of common prefix: this means that *if* there is some non-empty path in the base URL, that is not a prefix of the URL we are trying to shorten, then we are going to have to eliminate that by making sure there's a slash on the frond.

The "is not a prefix" thing in that previous sentence then needs clarification, because in general you don't have to consider anything after the final slash when deciding if something is a prefix, because of how relative paths are interpreted.

Anyway, this is what the last test case is trying to cover, and it shouldn't ever happen with the kinds of URL we will be passing around.

File: app/url_utils.py (hunk 2 above)
Comment (galund) [reply app/url_utils.py]: Quite: councils of perfection that result in complete inaction aren't the incremental agile change we're signed up to ;-) In this context, inconsistency is inevitable and good, as long as we're agreed about the direction of travel.

Comment (galund) [reply app/url_utils.py]: Incidentally I wanted to put this into dmutils, but I won't be doing that cos supporting python 2 is a PITA :-p

File: app/url_utils.py (hunk 1 above)
Comment (benvand) [reply app/url_utils.py]: but saying `c:my/weird/path` is the same as saying `c:/my/weird/path` so it doesn't matter if we always have a preceeding slash.
So we could just do this:
```
url = urlsplit(str(url))
path_string = urlunsplit(['', '', url.path, url.query, url.fragment])
return path_string if path_string.startswith('/') else '/' + path_string
```

Comment (benvand) [reply app/url_utils.py]: Grim. Tested this on my windows box at home
`c:my/weird/path is the same as saying c:/my/weird/path`
It's not. I stand corrected. I thought `c:` pointed to the mount point as an absolute.



Here's what I'm unsure of:
OK so the point of this story is to remove the hostname and port to prevent people injecting malicious urls into our db?

What do you mean by 'mounted somewhere other than `/`'?

If all the paths are going to be relative then why does it matter if they begin with a slash or not?

File: app/url_utils.py (hunk 2 above)
Comment (benvand) [reply app/url_utils.py]: Creating inconsistencies in the code base can make you feel like you're doing the _right_ thing, and you might well be doing the right thing, in this case you probably are, butmy problem is that it inconsistency _reduces confidence_ in the code. People will be confused by the multiple conventions and will wonder why that code had to be kept separately, or be tacked on, or not be properly integrated.

It doesn't reduce confidence in all the other code that follows the same patterns (no matter how wrong they might be); It reduces confidence in the different code, the inconsistent code.

I personally thing that we should be using class based views for all our views but if I started writing class based views it would look weird and out of place. And besides, that wouldn't achieve my goal of all views being class based. It would just mean we had 2 paradigms until I classed up all the views.

It's all about the easiest way for you to achieve your goal of not having a big utils file:
Say that it falls outside the scope of this ticket. (If this conversation was the only thing holding up this PR how annoying would that be?)
Highlight it to other members of the team to make sure they agree.
Then either:
 * Create a tech debt ticket and get someone else to do it in a separate PR all at once 🎉 
 * Do it yourself in a separate PR all at once
That way the PR will be accepted quickly because people will see it's just arefacotr and not have to go digging through a refactor for functional changes

Half staring a refactor and tagging it onto a bug fix PR doesn't help you achieve your goal of not having a massive utils file. It makes the PR harder to read, introduces confusion and reduces confidence.

I'm not trying to 💩  on the idea or anything like that. I think breaking the code into other files is a good idea. But I think that if we do it, we should do the whole thing. Every code base I've ever worked with has had a legacy of accepting PRs that introduce new styles and paradigms and conventions without evaluating closely enough the effort it will take to implement them fully. In this case the effort is low, but the cost of failing to complete the refactor is high (reduced confidence)

That's why I think consistency is important, even if what you're being consistent with isn't quite perfect.

That's the last I'll say on it, it's up to you if you want to change it.

File: app/url_utils.py (hunk 1 above)
Comment (samuelhwilliams) [reply app/url_utils.py]: I _think_ George means in the case that we change our search-api to be mounted at, say, www.digitalmarketplace.service.gov.uk/search-api/
